```
======================================================================
ERROR: test_selective_product_sync (corehq.apps.commtrack.tests.test_fixture.FixtureTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/dimagi/commcare-hq/corehq/apps/commtrack/tests/test_fixture.py", line 137, in test_selective_product_sync
    ).all(), key=lambda x: x.date)[-1]
TypeError: can't compare datetime.datetime to NoneType
----------------------------------------------------------------------
```

--- a/corehq/apps/commtrack/tests/util.py
+++ b/corehq/apps/commtrack/tests/util.py
@@ -1,5 +1,6 @@
 from django.test import TestCase
 from casexml.apps.case.tests import delete_all_cases, delete_all_xforms
+from casexml.apps.case.tests.util import delete_all_sync_logs
 from casexml.apps.case.xml import V2
 from casexml.apps.stock.models import StockReport, StockTransaction
 from casexml.apps.stock.const import COMMTRACK_REPORT_XMLNS
@@ -117,6 +118,8 @@ class CommTrackTest(TestCase):
         # might as well clean house before doing anything
         delete_all_xforms()
         delete_all_cases()
+        delete_all_sync_logs()
+
         StockReport.objects.all().delete()
         StockTransaction.objects.all().delete()
 

--- a/corehq/ex-submodules/casexml/apps/case/models.py
+++ b/corehq/ex-submodules/casexml/apps/case/models.py
@@ -655,11 +655,12 @@ class CommCareCase(SafeSaveDocument, IndexHoldingMixIn, ComputedDocumentMixin,
                                                               create_action)
         self.actions.append(create_action)
     
-    def update_from_case_update(self, case_update, xformdoc):
+    def update_from_case_update(self, case_update, xformdoc, other_forms=None):
+        other_forms = other_forms or {}
         if case_update.has_referrals():
             return self._legacy_update_from_case_update(case_update, xformdoc)
         else:
-            return self._new_update_from_case_update(case_update, xformdoc)
+            return self._new_update_from_case_update(case_update, xformdoc, other_forms)
 
     def _legacy_update_from_case_update(self, case_update, xformdoc):
         mod_date = parsing.string_to_datetime(case_update.modified_on_str) \
@@ -740,7 +741,7 @@ class CommCareCase(SafeSaveDocument, IndexHoldingMixIn, ComputedDocumentMixin,
         if case_update.user_id:     self.user_id = case_update.user_id
         if case_update.version:     self.version = case_update.version
 
-    def _new_update_from_case_update(self, case_update, xformdoc):
+    def _new_update_from_case_update(self, case_update, xformdoc, other_forms):
         assert not case_update.has_referrals()
 
         mod_date = parsing.string_to_datetime(case_update.modified_on_str) \
@@ -756,7 +757,9 @@ class CommCareCase(SafeSaveDocument, IndexHoldingMixIn, ComputedDocumentMixin,
                 )
                 self.actions.append(case_action)
 
-        self.rebuild(strict=False, xforms={xformdoc._id: xformdoc})
+        local_forms = {xformdoc._id: xformdoc}
+        local_forms.update(other_forms)
+        self.rebuild(strict=False, xforms=local_forms)
 
         # override any explicit properties from the update
         if self.modified_on is None or mod_date > self.modified_on:

--- a/corehq/ex-submodules/casexml/apps/case/tests/test_multimedia.py
+++ b/corehq/ex-submodules/casexml/apps/case/tests/test_multimedia.py
@@ -250,3 +250,4 @@ class CaseMultimediaTest(BaseCaseMultimediaTest):
         # this used to fail before we fixed http://manage.dimagi.com/default.asp?158373
         self._doSubmitUpdateWithMultimedia(new_attachments=['commcare_logo_file'], removes=[],
                                            sync_token=sync_log._id)
+        sync_log.delete()

--- a/corehq/ex-submodules/casexml/apps/case/tests/test_multimedia.py
+++ b/corehq/ex-submodules/casexml/apps/case/tests/test_multimedia.py
@@ -12,6 +12,8 @@ from casexml.apps.case.models import CommCareCase
 from casexml.apps.case import process_cases
 from casexml.apps.case.tests.util import delete_all_cases, delete_all_xforms
 from casexml.apps.case.xml import V2
+from casexml.apps.phone.models import SyncLog
+import couchforms
 from couchforms.models import XFormInstance, XFormDeprecated
 from couchforms.tests.testutils import post_xform_to_couch
 
@@ -75,19 +77,24 @@ class BaseCaseMultimediaTest(TestCase):
         with open(MEDIA_FILES[key], 'rb') as attach:
             return hashlib.md5(attach.read()).hexdigest()
 
-    def _do_submit(self, xml_data, dict_attachments):
+    def _do_submit(self, xml_data, dict_attachments, sync_token=None):
         """
         RequestFactory submitter - simulates direct submission to server directly (no need to call process case after fact)
         """
-        form = post_xform_to_couch(xml_data, dict_attachments)
-        form.domain = TEST_DOMAIN
+        sp = couchforms.SubmissionPost(
+            instance=xml_data,
+            domain=TEST_DOMAIN,
+            attachments=dict_attachments,
+            last_sync_token=sync_token,
+        )
+        response, xform, cases = sp.run()
         self.assertEqual(set(dict_attachments.keys()),
-                         set(form.attachments.keys()))
-        case, = process_cases(form)
+                         set(xform.attachments.keys()))
+        [case] = cases
         self.assertEqual(case.case_id, TEST_CASE_ID)
 
-    def _submit_and_verify(self, doc_id, xml_data, dict_attachments):
-        self._do_submit(xml_data, dict_attachments)
+    def _submit_and_verify(self, doc_id, xml_data, dict_attachments, sync_token=None):
+        self._do_submit(xml_data, dict_attachments, sync_token)
 
         time.sleep(2)
         form = XFormInstance.get(doc_id)
@@ -107,14 +114,15 @@ class BaseCaseMultimediaTest(TestCase):
         final_xml = self._formatXForm(CREATE_XFORM_ID, xml_data, attachment_block)
         self._submit_and_verify(CREATE_XFORM_ID, final_xml, dict_attachments)
 
-    def _doSubmitUpdateWithMultimedia(self, new_attachments=['commcare_logo_file', 'dimagi_logo_file'],
-                                      removes=['fruity_file']):
+    def _doSubmitUpdateWithMultimedia(self, new_attachments=None, removes=None, sync_token=None):
+        new_attachments = new_attachments if new_attachments is not None \
+            else ['commcare_logo_file', 'dimagi_logo_file']
+        removes = removes if removes is not None else ['fruity_file']
         attachment_block, dict_attachments = self._prepAttachments(new_attachments, removes=removes)
-
         raw_xform = self._getXFormString('multimedia_update.xml')
         doc_id = uuid.uuid4().hex
         final_xform = self._formatXForm(doc_id, raw_xform, attachment_block)
-        self._submit_and_verify(doc_id, final_xform, dict_attachments)
+        self._submit_and_verify(doc_id, final_xform, dict_attachments, sync_token)
 
 
 class CaseMultimediaTest(BaseCaseMultimediaTest):
@@ -221,8 +229,7 @@ class CaseMultimediaTest(BaseCaseMultimediaTest):
 
     def testAttachInUpdate(self, new_attachments=['commcare_logo_file', 'dimagi_logo_file']):
         self.testAttachInCreate()
-        removes = []
-        self._doSubmitUpdateWithMultimedia(new_attachments=new_attachments, removes=removes)
+        self._doSubmitUpdateWithMultimedia(new_attachments=new_attachments, removes=[])
 
         case = CommCareCase.get(TEST_CASE_ID)
         #1 plus the 2 we had
@@ -235,3 +242,11 @@ class CaseMultimediaTest(BaseCaseMultimediaTest):
         for attach_name in new_attachments:
             self.assertTrue(attach_name in case.case_attachments)
             self.assertEqual(self._calc_file_hash(attach_name), hashlib.md5(case.get_attachment(attach_name)).hexdigest())
+
+    def test_sync_log_invalidation_bug(self):
+        sync_log = SyncLog(user_id='6dac4940-913e-11e0-9d4b-005056aa7fb5')
+        sync_log.save()
+        self.testAttachInCreate()
+        # this used to fail before we fixed http://manage.dimagi.com/default.asp?158373
+        self._doSubmitUpdateWithMultimedia(new_attachments=['commcare_logo_file'], removes=[],
+                                           sync_token=sync_log._id)

--- a/corehq/ex-submodules/casexml/apps/case/xform.py
+++ b/corehq/ex-submodules/casexml/apps/case/xform.py
@@ -127,13 +127,14 @@ class CaseDbCache(object):
     to the database. Also provides some type checking safety.
     """
     def __init__(self, domain=None, strip_history=False, deleted_ok=False,
-                 lock=False, wrap=True, initial=None):
+                 lock=False, wrap=True, initial=None, xform=None):
         if initial:
             self.cache = {case['_id']: case for case in initial}
         else:
             self.cache = {}
 
         self.domain = domain
+        self.xform = xform
         self.strip_history = strip_history
         self.deleted_ok = deleted_ok
         self.lock = lock
@@ -223,6 +224,16 @@ class CaseDbCache(object):
     def clear_changed(self):
         self._changed = set()
 
+    def get_cached_forms(self):
+        """
+        Get any in-memory forms being processed.
+        """
+        # this will currently be at-most one form, though could be extended in the future
+        # abstracting the form cache inside this object seems useful.
+        if self.xform:
+            return {self.xform._id: self.xform}
+        return {}
+
 
 def get_and_check_xform_domain(xform):
     try:
@@ -288,14 +299,13 @@ def _get_or_update_model(case_update, xform, case_db):
     Gets or updates an existing case, based on a block of data in a
     submitted form.  Doesn't save anything.
     """
-
     case = case_db.get(case_update.id)
 
     if case is None:
         case = CommCareCase.from_case_update(case_update, xform)
         return case
     else:
-        case.update_from_case_update(case_update, xform)
+        case.update_from_case_update(case_update, xform, case_db.get_cached_forms())
         return case
 
 

--- a/corehq/ex-submodules/couchforms/util.py
+++ b/corehq/ex-submodules/couchforms/util.py
@@ -427,7 +427,7 @@ class SubmissionPost(object):
                 instance = xforms[0]
                 if instance.doc_type == "XFormInstance":
                     domain = get_and_check_xform_domain(instance)
-                    with CaseDbCache(domain=domain, lock=True, deleted_ok=True) as case_db:
+                    with CaseDbCache(domain=domain, lock=True, deleted_ok=True, xform=instance) as case_db:
                         try:
                             process_cases_with_casedb(instance, case_db)
                             process_stock(instance, case_db)

